### DISCUSSION TITLE
Automatic Evaluation with Simulated Annotator

### DISCUSSION THREAD

**Discussion by Zrrr1997** [Project-MONAI/MONAILabel#1188] [General]
Hi,

thanks for this wonderful project! I was wondering whether it is possible to automatically evaluate trained models by simulating the user during the testing stage. Most works in interactive segmentation start from a random initialization of clicks and then iteratively add clicks in the **center of largest connected error component**, similar to the training handlers in MONAI Label.

Is there a tutorial or a piece of code that I missed out regarding doing this during the **evaluation**? I would like to see how a trained model, e.g., DeepEdit, performs on the whole dataset, for which it would be impractical to conduct a user study with real annotators. A simulated annotator would also give a good indication of how the performance increases with the number of interactions (i.e. IoU@NoC). 
 
 Thanks very much in advance and keep up the great work!
 
 Best,
 Zdravko

**Comment by tangy5** [comment]
As far as I know, MONAI Label performs evaluation on validation set or auto partitioned set when conducting training. And active learning or scoring process such as epistemic scoring can score each unlabeled data with an uncertainty measure. Except these, there is no further "simulated annotator", maybe @diazandr3s can help more here about DeepEdit.
But I think this is a good point, **evaluate how good a current trained model is on entire dataset**  might be very useful when the interactive annotation performs. If currently no perfect solution, you are very welcome to contribute the "simulated annotator", the open-sourced project is looking for ideas/discussion/feedback/contribution. 

Thank you for the discussion.

**Reply by Zrrr1997** [reply]
Interesting, I think that a simulated user would be a very nice contribution since it would enable us to automatically verify and **compare methods on an equal footing**. Most works I have read go for using simply the center of the object as an initialization foreground click and then just sample from the **center of the largest connected error component**, which could be either the background or foreground ([ITIS: BMVC 2018](http://bmvc2018.org/contents/papers/0652.pdf), [RITM: ICIP 2022](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=9897365)). 

I dug around the tutorials and actually found the way **DeepEdit** does this: It uses the Chamfer Distance Transform (CDT) over the discrepancy regions (over- and undersegmentations) and the click is sampled from the pseudo-probability of the logits of the network (in this case DeepEdit based on DynUNet or Swin UNetr). So the clicks are more randomized than just using the center of the largest error. I have only seen this method used in [Sakinis et al. (2019)](https://arxiv.org/pdf/1903.08205.pdf), which seems to be the base idea behind DeepEdit. 

The code for this is [here](https://github.com/Project-MONAI/MONAILabel/blob/a2953b626eee1d48681e1aa3facab84c43b534af/sample-apps/radiology/lib/trainers/deepedit.py#L83), but you really need to dive into the source code to understand what exactly these transformations are doing. 

I think this will do for now, but I can gladly add a contribution regarding further annotator simulations, e.g., [the center of largest error](http://bmvc2018.org/contents/papers/0652.pdf) or [extreme points](https://openaccess.thecvf.com/content_cvpr_2018/papers/Maninis_Deep_Extreme_Cut_CVPR_2018_paper.pdf) or [internal margin points](https://www.sciencedirect.com/science/article/abs/pii/S1361841521001481) (and many others).

**Reply by SachidanandAlle** [reply]
FYI... we do that for Deepgrow models..
